Convert Payment Sheet RecylerViewHolders to Compose

skyler-stripe opened the pull request:
# Summary
<!-- Simple summary of what was changed. -->
This PR converts our payment method selectors to compose.

# Motivation
<!-- Why are you making this change? If it's for fixing a bug, if possible, please include a code snippet or example project that demonstrates the issue. -->
1. I like compose better. (I'm still learning so I expect I'm not doing a few things optimally) 
2. We need to theme these recylers eventually and making them compose will make it easier. Note that I didn't convert the whole recycler. There's a lot of logic to move over and I don't think there would be benefits to that quite yet. 

# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [ ] Added tests
- [x] Modified tests
- [x] Manually verified

I checked accessibility too. It should match 1:1 from master. Test various states of enable/disable with various payment methods and screen sizes. I will share these screenshots with JJ and fix any issues in a follow up PR that includes a few other minor fixes: https://www.figma.com/file/1Z3pj7oLK19jRSedqSufAx/Appearance-customization?node-id=40%3A1265

# Screenshots
| Before  | After |
| ------------- | ------------- |
|![AddCardBefore](https://user-images.githubusercontent.com/89166418/155437244-7b1d54ef-9367-4f6a-a3cb-d733e509d2c9.png)| ![AddCardAfter](https://user-images.githubusercontent.com/89166418/155437285-77e5488c-be15-4670-bc7f-9a94d70cde09.png)|
|![SelectorBefore](https://user-images.githubusercontent.com/89166418/155437367-b60553c0-eda7-4952-b43b-88270e3dc4ea.png)|![SelectorAfter](https://user-images.githubusercontent.com/89166418/155437390-d899c1cd-d17b-4ea5-a48a-438d580bb531.png)|
|![SelectorBeforeEditing](https://user-images.githubusercontent.com/89166418/155437469-1097249a-ea69-48b7-ba7a-ae3fc9b843ab.png)|![SelectorAfterEditing](https://user-images.githubusercontent.com/89166418/155437491-bab13a91-107a-467e-a54b-e7c7172856ca.png)|
|![tabletBefore](https://user-images.githubusercontent.com/89166418/155437593-630b1490-5efc-401e-b9da-94eafa11d65f.png)|![tabletAfter](https://user-images.githubusercontent.com/89166418/155437604-fdd464ee-40ae-4533-99ee-d43f72af55cd.png)|
|![addDarkBefore](https://user-images.githubusercontent.com/89166418/155619330-bc9f22d8-2802-41a9-acad-8b0d73672dff.png)|![addDarkAfter](https://user-images.githubusercontent.com/89166418/155619357-aa66b52c-ada2-4bcc-bb9e-73695cb9a0bc.png)
|![selectorDarkBefore](https://user-images.githubusercontent.com/89166418/155619362-fc6e618e-a9bc-4937-9c89-deb4072f5b56.png)|![selectorDarkAfter](https://user-images.githubusercontent.com/89166418/155619376-9c47682e-0223-498f-9010-9d5fada73987.png)|

Edge Cases 
| Long Text | Tablet with several methods |
| ------------- | ------------- |
|![longtext](https://user-images.githubusercontent.com/89166418/155584329-fa4aff75-1efc-47f4-8d26-0c60ebd5f65b.png)|![tablet2](https://user-images.githubusercontent.com/89166418/155620473-ba9fead5-a1ba-4e0e-8616-ebcf9ed8e654.png)|

# Changelog
This should be an invisible change to end users, I don't think it belongs in the change log.

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+        colorResource(R.color.stripe_paymentsheet_elements_background_disabled)
+    }
+
+    Card(
```
When we get to adding the Returning user selector, hopefully we can reuse this "Card" component

skyler-stripe replied:
> This is a great first iteration to just refactor not rewrite! I really like the approach here!
> 
> I am curious if you have follow-up PRs related to these changes planned? For instance it seems like the Card element is not reused between the two selectors, and perhaps that is due to current design (it looks like the color selector is different between the two). It seems like this is something that we should resolve, but not necessarily in this review.
> 
> In the past I have tracked things like this in the PR description, JIRA, and a paper doc. I don't have a strong preference, whatever you prefer.

Replying to both. My ideal end goal (will make a tech debt for and circle back after some customization work) is to have a single compose view that replaces the whole recycler view that works for both selectors. We should strive eventually to use the same views for both recyclers. For now I think it's fine to keep them different since their UIs have some nuance that is different.

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+        Column {
+            Image(
+                painter = painterResource(iconRes),
+                contentDescription = "",
```
Just want to double check that the image content description is not needed here for accessibility?

skyler-stripe replied:
Yup leaving it blank makes the reader not read out "ICON" lol.

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+                    .padding(top = 12.dp, start = 12.dp)
+            )
+            val color = colorResource(R.color.stripe_paymentsheet_title_text)
+            Text(
```
nit: Making this a composable named LpmSelectorText might improve readability a little.

skyler-stripe replied:
done

michelleb-stripe commented:
Can you post the dark mode screenshots as well, I think there might be a couple of issues with the Card background color?

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentOptionsAdapter.kt:
```diff
@@ -428,6 +448,145 @@ internal class PaymentOptionsAdapter(
+        colorResource(id = R.color.stripe_paymentsheet_elements_background_disabled)
+    }
+
+    ConstraintLayout(
```
nit: Could you add a comment on why ConstraintLayout is required, I think there are some optimization if you use Columns and Rows where possible.

skyler-stripe replied:
done

michelleb-stripe commented:
This is a great first iteration to just refactor not rewrite!  I really like the approach here!

I am curious if you have follow-up PRs related to these changes planned?  For instance it seems like the Card element is not reused between the two selectors, and perhaps that is due to current design.  It seems like this is something that we should resolve, but not necessarily in this review.

In the past I have tracked things like this in the PR description, JIRA, and a paper doc.  I don't have a strong preference, whatever you prefer.

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+        modifier = Modifier
+            .height(60.dp)
+            .width(viewWidth)
+            .padding(start = 6.dp, end = 6.dp)
```
```suggestion
            .padding(horizontal = 6.dp)
```

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+                painter = painterResource(iconRes),
+                contentDescription = "",
+                modifier = Modifier
+                    .padding(top = 12.dp, start = 12.dp)
```
nit: It seems like the start of 12.dp is used in a couple of places, might be better as a static constant

skyler-stripe replied:
done

michelleb-stripe commented:
Another test to try out is the card view when the description of the LPM doesn't fit in the card.  A good case of this is SEPA (delayed EUR LPM) when displayed in the German language.

<img width="356" alt="image" src="https://user-images.githubusercontent.com/77996191/155546852-e911d8ae-1804-4840-8a98-9698982ae31b.png">

michelleb-stripe commented:
Did you have a chance to test out the Don't keep activities with these changes.  Keep in mind the existing issues documented in MOBILESDK-252.

michelleb-stripe commented on paymentsheet/api/paymentsheet.api:
```diff
@@ -1,3 +1,7 @@
+public final class com/stripe/android/paymentsheet/AddPaymentMethodsAdapterKt {
```
Can these compose functions be made internal?

skyler-stripe replied:
Yes. That was an oversight.

michelleb-stripe commented on paymentsheet/build.gradle:
```diff
@@ -21,6 +21,7 @@ if (System.getenv("JITPACK")) {
 dependencies {
     implementation project(":payments-core")
     implementation project(':payments-ui-core')
+    implementation "androidx.constraintlayout:constraintlayout-compose:1.0.0"
```
Can we use the libraries.androidx.* for this dependency.  See this [PR](https://github.com/stripe/stripe-android/pull/4601/files) for reference.

jameswoo-stripe replied:
Should this also point to `$composeVersion`?

skyler-stripe replied:
Added a constant in case we want to use it in other places, but we have to pull it in like other compose libs.

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -27,36 +43,21 @@ internal class AddPaymentMethodsAdapter(
     override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AddPaymentMethodViewHolder {
-        return AddPaymentMethodViewHolder(parent)
-            .apply {
-                val targetWidth = parent.measuredWidth - parent.paddingStart - parent.paddingEnd
```
I really like this change to do the calculation in a separate function.  Really helps readability!

michelleb-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/PaymentOptionsAdapter.kt:
```diff
@@ -428,6 +448,145 @@ internal class PaymentOptionsAdapter(
+                Image(
+                    alpha = if (isEnabled) 1.0F else 0.6F,
+                    painter = painterResource(iconRes),
+                    contentDescription = "",
```
Just double checking that this isn't needed for accessiblity

skyler-stripe replied:
same as above. The empty string prevents the reader from mentioning the image.

brnunes-stripe replied:
Nit: It should be null instead of empty.

> Some visual elements are purely decorative and you might not want to communicate them to the user. When you set the contentDescription parameter to null, you indicate to the Android framework that this element does not have associated actions or state.

https://developer.android.com/jetpack/compose/accessibility

michelleb-stripe reviewed (dismissed):
Really cool to see more things being converted compose, and I like the refactoring and iterative approach over a re-write.  Nicely done!

jameswoo-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +75,122 @@ internal class AddPaymentMethodsAdapter(
+        private fun calculateViewWidth(parent: ViewGroup, numberOfPaymentMethods: Int): Dp {
+            val targetWidth = parent.measuredWidth - parent.paddingStart - parent.paddingEnd
+            val screenDensity = parent.context.resources.displayMetrics.density
+            val minItemWidth = 100 * screenDensity + 6.0f + 6.0f
```
Where did we get these numbers from? Can we describe these numbers so it is more readable?

skyler-stripe replied:
fixed this with some constants.

skyler-stripe commented:
@michelleb-stripe I think I addressed all of the comments. Added some additional screenshots too.

 I need to double check keep activities but I think I didn't regress anything.

brnunes-stripe commented on build.gradle:
```diff
@@ -67,6 +67,10 @@ ext {
+    coreKtxVersion = '1.7.0'
+    composeVersion = '1.1.0'
+    composeConstraintLayout = '1.0.0'
+    composeActivityVersion = '1.4.0'
```
`androidx.activity:activity-compose` is a submodule of the `androidx.activity` with the same version as `androidxActivityVersion`above: https://developer.android.com/jetpack/androidx/releases/activity
Where is `coreKtxVersion` used? That's the same as `androidxCoreVersion`.
Where is `composeVersion` used? That's the same as `androidxComposeVersion`.
`composeActivityVersion` seems to be the only new one needed, because it's a different version than the `androidxConstraintlayoutVersion` above: https://developer.android.com/jetpack/androidx/releases/constraintlayout
Let's follow the same pattern for naming, and make it `androidxConstraintlayoutComposeVersion`, list it above in lexicographical order.

skyler-stripe replied:
fixed. idk what happened when I merged.

brnunes-stripe commented on paymentsheet/src/main/java/com/stripe/android/paymentsheet/AddPaymentMethodsAdapter.kt:
```diff
@@ -74,42 +74,120 @@ internal class AddPaymentMethodsAdapter(
+            LpmSelectorText(
+                text = title,
+                isEnabled = isEnabled,
+                modifier = Modifier.padding(top = 6.dp, start = ADD_PM_DEFAULT_PADDING.dp)
```
Should it really be only top and start? Don't we want to also have a padding at the end?

skyler-stripe replied:
Currently in Master it looks like this:
![Screen Shot 2022-02-25 at 9 59 14 AM](https://user-images.githubusercontent.com/89166418/155766782-ad55daf8-7879-47a3-b218-df4cfd3a926a.png)

I'm going to sync with JJ and ask what to do in this situation, but not using end padding makes it look like my screenshot above which I think is better.

brnunes-stripe replied:
Maybe ellipsize for now? 
`maxLines = 1, overflow = TextOverflow.Ellipsis`

skyler-stripe replied:
That's better. 
![Screen Shot 2022-02-25 at 11 09 54 AM](https://user-images.githubusercontent.com/89166418/155773839-e5283961-bb7c-48df-8b86-287d3fa1325e.png)

brnunes-stripe commented on paymentsheet/src/test/java/com/stripe/android/paymentsheet/PaymentSheetAddPaymentMethodFragmentTest.kt:
```diff
@@ -97,7 +97,7 @@ internal class PaymentSheetAddPaymentMethodFragmentTest : PaymentSheetViewModelT
         createFragment(stripeIntent = paymentIntent) { _, viewBinding, _ ->
             val item = viewBinding.paymentMethodsRecycler.layoutManager!!.findViewByPosition(0)
-            assertThat(item!!.measuredWidth).isEqualTo(104)
+            assertThat(item!!.measuredWidth).isEqualTo(116)
```
Why did these values change?

skyler-stripe replied:
Previously we were measuring just the card part of the view. Now we have to measure the whole compose view for this test, which now includes (116-104) 12dp of padding
